pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotInstanceAssignmentRestletResource.java

```diff
@@ -0,0 +1,314 @@
+                .assignInstances(InstancePartitionsType.OFFLINE, instanceConfigs));
+          }
+        } catch (IllegalStateException e) {
+          throw new ControllerApplicationException(LOGGER, "Caught IllegalStateException", Response.Status.BAD_REQUEST,
+              e);
+        } catch (Exception e) {
+          throw new ControllerApplicationException(LOGGER, "Caught exception while calculating the instance partitions",
+              Response.Status.INTERNAL_SERVER_ERROR, e);
+        }
+      }
+    }
+    if (tableType != TableType.OFFLINE && instancePartitionsType != InstancePartitionsType.OFFLINE) {
+      TableConfig realtimeTableConfig = _resourceManager.getRealtimeTableConfig(tableName);
+      if (realtimeTableConfig != null) {
+        try {
+          InstanceAssignmentDriver instanceAssignmentDriver = new InstanceAssignmentDriver(realtimeTableConfig);
+          if (instancePartitionsType == InstancePartitionsType.CONSUMING || instancePartitionsType == null) {
+            if (InstanceAssignmentConfigUtils
+                .allowInstanceAssignment(realtimeTableConfig, InstancePartitionsType.CONSUMING)) {
+              instancePartitionsMap.put(InstancePartitionsType.CONSUMING,
+                  instanceAssignmentDriver.assignInstances(InstancePartitionsType.CONSUMING, instanceConfigs));
+            }
+          }
+          if (instancePartitionsType == InstancePartitionsType.COMPLETED || instancePartitionsType == null) {
+            if (InstanceAssignmentConfigUtils
+                .allowInstanceAssignment(realtimeTableConfig, InstancePartitionsType.COMPLETED)) {
+              instancePartitionsMap.put(InstancePartitionsType.COMPLETED,
+                  instanceAssignmentDriver.assignInstances(InstancePartitionsType.COMPLETED, instanceConfigs));
+            }
+          }
+        } catch (IllegalStateException e) {
+          throw new ControllerApplicationException(LOGGER, "Caught IllegalStateException", Response.Status.BAD_REQUEST,
+              e);
+        } catch (Exception e) {
+          throw new ControllerApplicationException(LOGGER, "Caught exception while calculating the instance partitions",
+              Response.Status.INTERNAL_SERVER_ERROR, e);
+        }
+      }
+    }
+
+    if (instancePartitionsMap.isEmpty()) {
+      throw new ControllerApplicationException(LOGGER, "Failed to find the instance assignment config",
+          Response.Status.NOT_FOUND);
+    }
+
+    if (!dryRun) {
+      for (InstancePartitions instancePartitions : instancePartitionsMap.values()) {
+        persistInstancePartitionsHelper(instancePartitions);
+      }
+    }
+
+    return instancePartitionsMap;
+  }
+
+  @PUT
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("/tables/instancePartitions")
+  @ApiOperation(value = "Create/update the instance partitions")
+  public InstancePartitions putInstancePartitions(String instancePartitionsStr) {
```

Comment:
Opening this up will allow users to create arbitrarily named instance partitions. Is this what we want? If we only want to allow 1 instance partitions for each table, we need to associate instance partitions with the table. (but schema is already doing this.. :) )

e.g. `PUT /tables/{tableName}/instancePartitions`